Add smartfan suspend resume control

zmelumian972 opened the pull request:
Currently the CE will immediately take control over hwmon devices

this fix will take the control from the CE during resume, and give control to the CE during suspend

this will ensure that during suspend the temperature monitoring done by the CE will still be live, even if the OS is on suspend, keeping the PC healthy

zmelumian972 commented:
yeah, I missed your routines, I will find them and fix it

as for your secondary concern - let me try to verify that

TsunamiMommy commented:
So check lines 2904, and 3008-3012. Those lines will have what you need. Though the smartfan global routine may need to be adjusted to fit your use case. Hopefully it gives you an idea of how to set it. Also lines 3008-3012 have the guarding routine you need to check for ecio and mmio h2ram.

TsunamiMommy commented:
Something else I noticed is that you're using register 0x13 which isn't the actual smart fan control bit. The control bit for smartfan is at offset `0x947` in the h2ram register space and only applies to the secondary superio with the ecio and h2ram flags. So your current code does something else. It's turning fan control on and off completely. That's not necessarily what you want to be doing. Instead what you want to be doing is setting `0x947` on the secondary superio to `0x0`1 when going to sleep and then setting it to `0x00` or `0x01` on resume based on the state of your pwmX_enable sysfs values. So that way, when the chip wakes back up, it'll be in the correct state.

zmelumian972 committed:
Add smartfan suspend resume control

on suspend, we pass control of the hwmon devices to the CE

on resume, we take control back

zmelumian972 commented:
Updated

added a new routine `it87_update_smartfan_bit`  based on the content of `it87_update_smartfan_global` but added the option to define the set/clear bit

applied it to `it87_update_smartfan_global` , based on the value of `all_auto`

added checks for mmio_h2ram and ecio_h2ram

and added check at on resume to disable smartfan only if there isn't any pwm_enable set

I could verify that fan controllers over sioaddr are unaffected, but I don't know how to check for isa

EDIT: you were right about the address, I misread the chips documentation

TsunamiMommy commented:
Yeah will look at it in a bit. There are actually quite a few things that I'm working on regarding these chips as to support them properly will require quite a bit more than this. Though this should (in theory) be a "good enough" holdover until then. Have you tested your changes to make sure that they function correctly?

TsunamiMommy commented:
Well, then it seems like your modifications need to read out the currently set pwm_enable values first and then set smartfan afterwards based on a known system state.

TsunamiMommy commented:
@frankcrawford This isn't anywhere near being merge ready. It should be marked as draft for now.

TsunamiMommy commented:
This should be closed. I've handled the problem already and this PR is deprecated. Though, I'm still evaluating the need for this routine.

frankcrawford commented:
@zmelumian972 I'm going to close this PR now, as it is no longer valid following the ones from @TsunamiMommy 

If there are any further issues, please let us know.

frankcrawford commented:
Closing.
